core/src/main/java/com/linecorp/armeria/common/RequestContext.java

```diff
@@ -597,26 +597,18 @@ default ContextAwareBlockingTaskExecutor makeContextAware(BlockingTaskExecutor e
      * Returns a {@link Callable} that makes sure the current {@link RequestContext} is set and then invokes
      * the input {@code callable}.
      */
-    default <T> Callable<T> makeContextAware(Callable<T> callable) {
+    default <T> ContextAwareCallable<T> makeContextAware(Callable<T> callable) {
```

Comment:
Changed the return type following the principle of "return the most specialized; accept the most generic".

Reply:
This seems like a breaking change even with API, not just ABI, something like `Callable aware = ctx.makeContextAware(notAware);` wouldn't compile anymore.

I guess callers would just need to stick with a type-cast until a major version bump could change the signature, or change the target milestone of the PR to v2.0.0

Reply:
I think you're right @anuraaga - this could [break binary compatibility](https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html#jls-13.4.15).

Let's change it back to play it safe?

Reply:
Thanks yeah it seems safer to do that.

Just as an idea, static methods on `RequestContext` would be an alternative to exposing new types

```
interface RequestContext {

static RequestContext from(Callable c)
static Callable unwrap(Callable c)
static RequestContext from(Runnable)
static Runnable unwrap(Runnable)
// etc
```

Naturally it seems like it could get unwieldy if adding even more methods, but if the expectation for that is low it could be viable, just an idea

Reply:
That's sound idea to structure this. I think we might actually just wait until we can break the binary compatibility. Type-casting isn't too bad and is something I was going to do anyway as the place where I need `context()` gives me only Runnable.

Reply:
It breaks binary compatibility. We had a similar discussion before. If a change works well with recompilation without modification of users' code, we allowed the change in minor upgrade.

ref: https://github.com/line/armeria/pull/4760
> Simply recompiling your code should be enough in most cases because
BlockingTaskExecutor is a ScheduledExecutorService.

Reply:
Hmm - I have a server framework that uses Armeria that I haven't updated in a while as no new features added. It was nice that Armeria could still be updated in apps using it, and `makeContextAware` is definitely used so will break. Didn't look at the link maybe it was already broken though 😅 Not a good thing to do in Java since usually you don't compile all your code.

Reply:
I can't say that we strictly follow semantic versioning. Agree on minimizing making breaking changes and maintaining API and ABI compatibility where possible.

That said, if we stick with semantic versioning, we'll have to keep putting things off and Armeria won't evolve easily until the next major version. 

It is necessary making the current implicit versioning rule clearer and specify what can happen in a minor upgrade. For example, in the case of Spring Boot, when they release a new version, the previously deprecated functions in two minor versions ago are deleted.

Reply:
Strictly following semantic versioning will make us get to Armeria 9.0 fairly soon. 😄 We already introduced various ABI-breaking changes similar to this in 1.x unfortunately. Let's keep this PR as it is. Let us have some internal discussion on how we could handle this in our future releases.